pkg/controllers/nodepool/readiness/controller.go

```diff
@@ -0,0 +1,120 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+)
+
+// Controller for the resource
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+}
+
+// NewController is a constructor
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodePool *v1beta1.NodePool) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "nodepool.status")
+	stored := nodePool.DeepCopy()
+	supportedNC := c.cloudProvider.GetSupportedNodeClasses()
+
+	if len(supportedNC) == 0 {
+		logger.Fatal("no supported node classes found for the cloud provider")
+	}
+	nodeClass, err := c.getNodeClass(ctx, nodePool, supportedNC)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		log.FromContext(ctx).WithValues("NodeClass", nodePool.Spec.Template.Spec.NodeClassRef.Name).Info("nodeClass not found")
+		nodePool.StatusConditions().SetFalse(v1beta1.ConditionTypeNodeClassReady, "UnresolvedNodeClass", "Unable to resolve nodeClass")
+	} else {
+		c.setReadyCondition(nodePool, nodeClass)
+	}
+	if !equality.Semantic.DeepEqual(stored, nodePool) {
+		if err = c.kubeClient.Status().Update(ctx, nodePool); client.IgnoreNotFound(err) != nil {
+			if errors.IsConflict(err) {
+				return reconcile.Result{Requeue: true}, nil
+			}
+			return reconcile.Result{}, err
+		}
+	}
+	return reconcile.Result{}, nil
+}
+func (c *Controller) getNodeClass(ctx context.Context, nodePool *v1beta1.NodePool, supportedNC []status.Object) (status.Object, error) {
+	nodeClass := supportedNC[0]
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodePool.Spec.Template.Spec.NodeClassRef.Name}, nodeClass); err != nil {
+		return nil, err
+	}
+	return nodeClass, nil
+}
+
+func (c *Controller) setReadyCondition(nodePool *v1beta1.NodePool, nodeClass status.Object) {
+	ready := nodeClass.StatusConditions().Get(status.ConditionReady)
+	if ready.IsUnknown() {
+		nodePool.StatusConditions().SetFalse(v1beta1.ConditionTypeNodeClassReady, "NodeClassReadinessUnknown", "Node Class Readiness Unknown")
+	} else if ready.IsFalse() {
+		nodePool.StatusConditions().SetFalse(v1beta1.ConditionTypeNodeClassReady, ready.Reason, ready.Message)
+	} else {
+		nodePool.StatusConditions().SetTrue(v1beta1.ConditionTypeNodeClassReady)
+	}
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	builder := controllerruntime.NewControllerManagedBy(m)
+	for _, nodeClass := range c.cloudProvider.GetSupportedNodeClasses() {
+		builder = builder.Watches(
+			nodeClass,
+			nodepool.NodeClassEventHandler(c.kubeClient),
+		)
+	}
+	return builder.
+		Named("nodepool.status").
```

Comment:
same here on matching file name
```suggestion
		Named("nodepool.readiness").
```